src/common/src/main/scala/osmesa/common/functions/osm/package.scala

```diff
@@ -499,93 +505,115 @@ package object osm {
     }
   }
 
-  private def connectSegments(segments: GenTraversable[Line])(implicit geometryFactory: GeometryFactory): GenTraversable[Line] =
-    connectSegments(segments.map(_.jtsGeom.getCoordinateSequence).map(s => new VirtualCoordinateSequence(Seq(s))))
+  private def connectSegments(segments: GenTraversable[jts.Geometry])(implicit geometryFactory: GeometryFactory): GenTraversable[LineString] =
+    segments
+      .map({ geom => if (classTag[jts.LineString].runtimeClass.isInstance(geom)) Some(geom.asInstanceOf[jts.LineString]) else None })
+      .flatten
+      .map(_.getCoordinateSequence)
+      .map(s => new VirtualCoordinateSequence(Seq(s)))
       .map(geometryFactory.createLineString)
-      .map(Line(_))
 
-  private def dissolveRings(rings: GenTraversable[Polygon]): (Seq[Polygon], Seq[Polygon]) = {
-    MultiPolygon(rings.toArray).union.asMultiPolygon match {
```

Comment:
The unary `union` is really important here; it looks like it got lost in the refactor.

I believe `MultiPolygonRelationReconstructionSpec` includes some sample geometries that rely on this behavior.

Reply:
Great catch :+1: